plutus-ledger-api/test-plugin/Spec/Value/TokenName.hs

```diff
@@ -0,0 +1,49 @@
+{-# LANGUAGE BlockArguments    #-}
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE TypeApplications  #-}
+{-# OPTIONS_GHC -fplugin PlutusTx.Plugin #-}
+
+module Spec.Value.TokenName where
+
+import Data.ByteString (ByteString)
+import PlutusCore (DefaultFun, DefaultUni, NamedDeBruijn, someValue)
+import PlutusLedgerApi.V1.Value (TokenName)
+import PlutusTx (CompiledCode, getPlcNoAnn)
+import PlutusTx.Plugin.IsString (RawBytes, Utf8Encoded)
+import PlutusTx.TH (compile)
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.HUnit (assertBool, testCase, (@?=))
+import UntypedPlutusCore (Program (_progTerm), Term (Constant))
+
+tests :: TestTree
+tests =
+  testGroup
+    "IsString TokenName"
+    [ test_TokenName_IsString_Utf8
+    , test_TokenName_IsString_Raw
+    ]
+
+test_TokenName_IsString_Utf8 :: TestTree
+test_TokenName_IsString_Utf8 = testCase "Utf8Encoded" do
+  term compiledLiteral @?= Constant () (someValue tokenName)
+ where
+  compiledLiteral :: CompiledCode (Utf8Encoded TokenName) =
+    $$(compile [||"Имя Токена"||])
+
+  tokenName :: ByteString -- UTF-8 encoded "Имя Токена"
+  tokenName =
+    "\208\152\208\188\209\143\32\
+    \\208\162\208\190\208\186\208\181\208\189\208\176"
+
+test_TokenName_IsString_Raw :: TestTree
+test_TokenName_IsString_Raw = testCase "Raw" do
+  let expectedTerm = Constant () (someValue ("\186\190" :: ByteString))
+  assertBool "compiledRawBytes == expectedTerm" $
+    term ($$(compile [||"\xBA\xBE" :: RawBytes TokenName||])) == expectedTerm
```

Comment:
This expression doesn't compile, `RawBytes` only supports a subset of UTF-8 code-points that are encoded with one byte, while code point И is encoded with [2 bytes](https://symbl.cc/en/0418/); This is by design: its not a good idea for users to use unicode codepoints to specify array of bytes. We expect users to use either decimal "\42" or hex "\xFF" byte syntax for this. 

GHC behaves differently: it truncates codepoints and warns users about it:
![image](https://github.com/user-attachments/assets/d6fd74f2-f827-49e9-b4c0-2cdc78e2379d)

I'd argue that compile time error is better than runtime truncation.

Reply:
This probably won't be needed anyway, at least for TokenName, because TokenName is supposed to be UTF-8, and it does make sense to use `RawBytes TokenName`.

But I think it's fine if not desirable for this to have the same behavior as GHC ByteString, for consistency.
So there are two behaviors:
- UTF-8
- Truncating, a la GHC ByteString, and `BuiltinByteString`

You are introducing a third behavior: same as above, except failing rather than truncating - that's one additional behavior to have to keep in mind which is imo unnecessary.

Reply:
> This probably won't be needed anyway, at least for TokenName, because TokenName is supposed to be UTF-8, and it doesn't make sense to use RawBytes TokenName.

Why do you say that `TokenName` is supposed to be UTF-8? Is it specified anywhere in the spec (I couldn't find it). 
I was operating on the assumption that users make a decision what encoding they want to use, if any.

Reply:
What I mean is each type can only have one `IsString` instance. If someone wants a different behavior they'd make a different newtype around `BuiltinByteString` and call it "MyTokenName" or whatever.

Reply:
Ok, thanks for the clarification.

Here is what you're suggesting:
| ****                         | **GHC**   | **Plinth**        |
|:----------------------------:|:---------:|:-----------------:|
| **Def Behaviour**            | UTF-8     | UTF-8             |
| **Values out of byte range** | Truncated | Truncated |

I am suggesting to replace the truncating behaviour in Plinth:
| ****                         | **GHC**   | **Plinth**        |
|:----------------------------:|:---------:|:-----------------:|
| **Def Behaviour**            | UTF-8     | UTF-8             |
| **Values out of byte range** | Truncated | Compilation error |
 
Even if users expect truncation (or wrapping around) they won't be surprised in runtime, as the compilation error message won't let it compile.

Reply:
Regardless of whether we truncate or fail-fast, here's what I'm suggesting in terms of implementation:

- Only a single newtype is needed: `newtype BuiltinByteStringUtf8 = BuiltinByteStringUtf8 { unBuiltinByteStringUtf8 :: BuiltinByteString }`
- Its `IsString` can be defined as `fromString s = BuiltinByteStringUtf8 . Builtins.encodeUtf8 . fromString`
- The plugin _can_ compile this `fromString` definition, so all we need to do in the plugin is to allow it to proceed with the normal compilation when it encounters `fromString @BuiltinByteStringUtf8`, rather than giving up.

If you want to opt for fail-fast rather than truncating, then modify the `fromString BuiltinByteString` to make it fail-fast.

wdyt?

Reply:
> so all we need to do in the plugin is to allow it to proceed with the normal compilation when it encounters fromString @BuiltinByteStringUtf8, rather than giving up.

I've spend most of the workday trying to make this "fall through the `fromString` compilation" work but didn't succeed :(

The problem looks like this:
```
<no location info>: error:
    GHC Core to PLC plugin: Error: Unsupported feature: Type constructor: GHC.Prim.Char#
Context: Compiling type: GHC.Prim.Char#
Context: Compiling data constructor type: GHC.Types.C#
Context: Compiling type: GHC.Types.Char
Context: Compiling type: GHC.Base.String
Context: Compiling type: GHC.Base.String -> a
Context: Compiling type: Data.String.IsString a
Context: Compiling type: Data.String.IsString a => GHC.Base.String -> a
Context: Compiling type: forall a. Data.String.IsString a => GHC.Base.String -> a
Context: Compiling definition of: Data.String.fromString
Context: Compiling expr: Data.String.fromString
Context: Compiling expr: Data.String.fromString @PlutusLedgerApi.V1.Value.TokenName
Context: Compiling expr: Data.String.fromString @PlutusLedgerApi.V1.Value.TokenName PlutusLedgerApi.V1.Value.$fIsStringTokenName
Context: Compiling expr: 
  Data.String.fromString
    @PlutusLedgerApi.V1.Value.TokenName
    PlutusLedgerApi.V1.Value.$fIsStringTokenName
    (GHC.CString.unpackCStringUtf8# 
       "\\208\\152\\208\\188\\209\\143 \\208\\162\\208\\190\\208\\186\\208\\181\\208\\189\\208\\176"#)
```
My understanding is that the definition of `fromString` itself is not compilable...

GHC sometimes won't inline the `fromString` no matter how hard I try to convince it.

Reply:
Assuming that proceeding with the compilation won't work, here are the options left on the table that I can identify:
1. (currently implemented in the PR) Plugin intercepts `fromString @Utf8Encoded` / `fromString @RawBytes` and users chose which one to use by prepending `utf8Encoded "Токен"` / `rawBytes "\xFF"` to string literals.
2. No `instance IsString TokenName` is defined thus requiring users to apply the newtype constructor (`TokenName`  / `CurrenctySymbol` / etc.) to a `BuiltinByteString` value directly. Users will still have to obtain the compile-time-UTF8-encoded `BuiltinByteString` but that could be done with the newtype wrapper around it + a hook in the plugin, e.g. like you suggest the `newtype BuiltinByteStringUtf8`. 

In either case Plugin will have to know what encoding is needed by looking at the [new]type to which `fromString` is applied.

Reply:
This is compiling `fromString @TokenName`. I'm no longer suggesting making `tn = "foo" :: TokenName` work, which seems complicated. Instead we can make users wrap/unwrap like you initially suggested, i.e., `tn = TokenName (unBuiltinByteStringUtf8 "foo")`.

Reply:
You can also intercept `fromString @BuiltinByteStringUtf8` and compile it in special ways, just like how you intercept `fromString @Utf8Encoded` and `fromString @RawBytes`. I think we just need one newtype: why do we need `RawBytes` if we already have `BuiltinByteString` and they behave the same?

However, I _think_ this is slightly more complex than making it proceed normally, since you need to worry about `fromString` being specialized or inlined.

Reply:
>  I'm no longer suggesting making tn = "foo" :: TokenName work

Yeah, it was @effectfully who suggested that we should stop intercepting `IsString.fromString` and instead rely on its inlining + internal opaque functions that are also intercepted by the plugin. Unfortunately I couldn't make it work.

> Instead we can make users wrap/unwrap like you initially suggested

🤝 

> You can also intercept fromString @BuiltinByteStringUtf8 and compile it in special ways, just like how you intercept `fromString @Utf8Encoded` and `fromString @RawBytes`. 

Agree, this is possible.

> I think we just need one newtype: why do we need `RawBytes` if we already have `BuiltinByteString` and they behave the same?

Say we have 
```haskell
newtype Foo = MkFoo BuiltinByteString
```
You're right that essentially `"..." :: RawBytes Foo` has the same behaviour as `MkFoo "..."`: in both cases the UTF-8 encoding done by GHC will be un-done in compile time and the same `Foo` value will be created.  The former approach allows "deriving via" 
```haskell
  deriving (IsString) via (Utf8Encode Foo)
```
but it doesn't seem to have a significant difference. 
You convinced me that for such newtypes that wrap `BuiltinByteString` but expect UTF-8 the direct approach of applying newtype constructor to `BultinByteString` value is a better option.

Reply:
> Yeah, it was @effectfully who suggested that we should stop intercepting IsString.fromString and instead rely on its inlining + internal opaque functions that are also intercepted by the plugin.

Well, intercepting `fromString` is not necessary for compilation, but it is necessary in order to produce a nice error message: "fromString Foo is not supported".

Reply:
> Yeah, it was @effectfully who suggested that we should stop intercepting IsString.fromString and instead rely on its inlining + internal opaque functions that are also intercepted by the plugin. Unfortunately I couldn't make it work.

No, I suggested something stupid: stop intercepting `IsString.fromString` and _not_ rely on its inlining. I thought that would still compile. But it of course can't because `String` is `[Char]` and we can't compile `Char`.

We can probably intercept `fromString` and massage it into something more manageable? Like `fromText :: Text -> str` defined specifically for that purpose.

Or maybe we can generally compile `[Char]` as `Text` somehow? That won't work when `[a]` is instantiated as `[Char]`, but maybe that's fine.